### DISCUSSION TITLE
Use noe-tree api commands outside of  config mappings?

### DISCUSSION THREAD

**Discussion by JoseConseco** [nvim-neo-tree/neo-tree.nvim#763] [Q&A] [answered]
He, I'm trying to write  [Hydra](https://github.com/anuvyklack/hydra.nvim) config for neo-tree (it allows u to draw popup with keys - legend, and execute command on key press).  It seems neo-tree expect some variables to be passed to commands eg. state, etc. , which are automatically set in  netTree mappings function.  
But in my case I want to call NeoTree commands outside of config.  I tried
```lua
local state = require("neo-tree.sources.manager").get_state("filesystem")
```
But then using:
```lua
{ "n", function() state.commands.add(state) end,                 { silent = true } },
```
In Hydra gives error about state missing config.  And other commands require passing multiple variables...
In NvimTree we get access to :  api.new   - which works ok outside of config. 
What is best way to  call NeoTree functions outside of config mappings, without having to pass all the variables (or how do  I generate these variables properly) ?

**Comment by miversen33** [comment]
[Take a peek at my dotfiles](https://github.com/miversen33/miversen-dotfiles/blob/774552e47c8dd6da45a05040c0e299dda0601219/editors/nvim/lua/plugins/keymaps.lua#L90-L129), I do this exact thing. Consider using the available [vim commands](https://github.com/nvim-neo-tree/neo-tree.nvim#the-neotree-command) for this, that is what they are meant for.

```lua
{"f",      cmd 'Neotree filesystem reveal right', {desc = "Opens Neotree File Explorer", silent = true}},
```

**Comment by JoseConseco** [comment]
Thx  @miversen33  but what it try to do is to add individual NeoTree commands to  hydra - not just opening the NeoTree buffer. My idea is to show NeoTree commands popup when NeoTree buffer is focused.   [Here](https://github.com/JoseConseco/nvim_config/blob/59eb67e99db75a95fb8d0763a25a849a8e1992a7/lua/nvimTree/init.lua#L76) is my current setup for NvimTree.

**Reply by miversen33** [reply]
I'm not seeing the issue here. You can set a "BufEnter" autocommand that is triggered when the Neo-tree buffer is focused. When it is, you can then display your hydra. I'm not seeing what that has to do with the neo-tree api? Am I missing something?

**Reply by JoseConseco** [reply]
Here is what I try to make (example based on NvimTree) - popup is only shown when NvimTree buffer is active:
![image](https://user-images.githubusercontent.com/13521338/220399707-4bb73595-c5d6-4084-8435-8590a563b283.png)
Code for NeoTree:
```lua
local api = require"neo-tree.sources.filesystem.commands"  -- does not work reliably....
local state = require("neo-tree.sources.manager").get_state("filesystem")
local common = require("neo-tree.sources.common.commands")

local Hydra = require "hydra"
local function spawn_nvim_tree_hydra()
  nvim_tree_hydra = Hydra {
          name = "NeoTree",
          hint = hint,
          config = {
              color = "pink",
              invoke_on_body = true,
              buffer = 0, -- only for active buffer
              hint = {
                  position = "bottom",
                  border = "rounded",
              },
          },
          mode = "n",
          body = "H",
          heads = {
              { "w", "",     { silent = true } },
              { "c", "",     { silent = true } },
              { "/", function() api.filter_as_you_type(state) end,         { silent = true } },
              { "y", function() api.copy_to_clipboard(state) end,              { silent = true } },
              { "x", function() api.cut_to_clipboard(state) end,                    { exit = true, silent = true } },
              { "p", function() api.paste_from_clipboard(state) end,                  { exit = true, silent = true } },
              { "r", function() api.rename(state) end,                 { silent = true } },
              { "d", function() api.delete(state) end,                 { silent = true } },
              { "n", function() state.commands.add(state) end,                 { silent = true } },
              { ".", function() api.toggle_hidden(state) end, { silent = true } },
              { "?", "",          { silent = true } },
          },
      }
  nvim_tree_hydra:activate()
end      
```
The problem is using :  state.commands.add(state) and other command often fail with errors, for example in this case  about state.config being nil.  Slightly different  error is shown for filter_as_you_type, etc.   How do I fix it?

**Reply by miversen33** [reply]
Ahh I am tracking now. I will circle back to this after work. Its not impossible, you just need a way to get state. I don't remember off the top of my head how to do this.

**Reply by JoseConseco** [reply]
@miversen33 any update? Would be great to get the correct state outside of NeoTree config.mappings

**Reply by miversen33** [reply]
@JoseConseco I apologize, I got a bit side tracked last night.

So after some digging, unfortunately it appears there is no easy way to do this. This is because the commands that are called are generated in a local function with the state already available to them. Meaning there is no _state_ lookup that happens on command call, the command just automatically has the state available to it.

[Source](https://github.com/nvim-neo-tree/neo-tree.nvim/blob/08d7cc91b429c1a415bfa87710cda6665ea39213/lua/neo-tree/ui/renderer.lua#L777-L799)

I am not completely sure how to "get" the state from outside Neo-tree and I would be very hesitant to recommend any way to do this as that would be an undocumented feature (meaning it is subject to change).

Maybe a better thing to do is open an issue requesting better access to this part of the functionality? @cseickel thoughts on dynamically overriding the keymappings in neo-tree?

**Reply by JoseConseco** [reply]
Thx!  made it work with cseickel way.

**Comment by cseickel** [comment, marked as answer]
On the one hand, you already have the state part figured out with:
```lua
local state = require("neo-tree.sources.manager").get_state("filesystem")
```
which will work fine for a sidebar style file system tree.

The extra config argument that is added can just be an empty table if you don't want to use any configuration. That means all you need to do to make the example work is to append 
```
state.config = {}
```

You should really do that on every command invocation instead of treating state like a global variable. I would create a function like this:

```lua
local build_command = function(command_name)
  local fn = function()
    local cmds = require("neo-tree.sources.filesystem.commands")
    local state = require("neo-tree.sources.manager").get_state("filesystem")
    state.config = {}
    cmds[command_name](state)
  end
  return fn
end

local Hydra = require "hydra"
local function spawn_nvim_tree_hydra()
  nvim_tree_hydra = Hydra {
          name = "NeoTree",
          hint = hint,
          config = {
              color = "pink",
              invoke_on_body = true,
              buffer = 0, -- only for active buffer
              hint = {
                  position = "bottom",
                  border = "rounded",
              },
          },
          mode = "n",
          body = "H",
          heads = {
              { "w", "",     { silent = true } },
              { "c", "",     { silent = true } },
              { "/", build_command("filter"),         { silent = true } },
              { "y", build_command("copy_to_clipboard"),              { silent = true } },
              ...
          },
      }
  nvim_tree_hydra:activate()
end  
```

**That code is completely untested.** It may have bugs, but you get the idea.

**Reply by JoseConseco** [reply]
Thanks.  I made it work with simplified build_command: 
```lua
local build_command = function(command_name)
  local cmds = require("neo-tree.sources.filesystem.commands")
  local state = require("neo-tree.sources.manager").get_state("filesystem")
  -- state.config = {}
  cmds[command_name](state)
end  
```
State.config = {} does not seen necessary with this.  And there was no need to wrap it in additional inner `fn`.
I guess defining  state outside hydra did not work, since it was initialized only once - and its state did not get updated on  function call...

**Reply by cseickel** [reply]
The only issue you might have with this is that you could get the `state.config` value from running a prior command through the normal means. It's meant to be overwritten on each command invocation with the config for that specific command. I would recommend keeping that `state.config = {}` line to prevent any conflicts from popping up in the future.
